WIP AccountingApp

readevalprint opened the pull request:
The initial idea is that there are `Transactions` that describe the high level details: from, to, how much, exchange rate and also `TransactionUpdates` which shows the state of the `Transaction` over time.

> Jorge edit: close #13, close #14

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+        string reference;
+
+        uint period;
+        // Back reference the states?
```
We have to see what is cheaper gas wise. Having a `mapping (uint => TransactionAction)` could turn out to be cheaper

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+pragma solidity ^0.4.8;                                                         
+
+contract AccountingApp {
```
Need to add the Application subclass and conform to needed methods

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+pragma solidity ^0.4.8;                                                         
```
We are using 0.4.11

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+        address baseToken;
+        address baseValue;
+        
+        address from;
```
Either from or to are always the DAO address. We could save some gas just saving a `senderOrRecipient` address field (or a better name) and then a boolean that indicates what is the transaction direction.

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+        
+    
+    struct TransactionAction {
+        uint transactionId;
```
If we end up doing the mapping thing explained above, we can remove the id here and save 1 sstore per state change :)

izqui commented on contracts/kernel/organs/AccountingApp.sol:
```diff
@@ -0,0 +1,98 @@
+        TransactionState state;
+        string reason;
+        address actor;
+        uint timestamp; 
```
uint64 are good enough for timestamps and save gas :)

readevalprint commented:
This is the current state machine, any feedback?
<img width="896" alt="screen shot 2017-07-07 at 11 28 42 am" src="https://user-images.githubusercontent.com/118430/27957585-5c154a24-631f-11e7-9384-fbe4397101a1.png">

izqui commented:
Makes sense

I like the idea of spliting sending a transaction in two actions, send for a approval and 'execute'. This will allow us to have different bylaws for both actions.

readevalprint commented:
Updated: added flow if the sender is an admin
 ![image](https://user-images.githubusercontent.com/118430/27960280-ff34e736-632a-11e7-9f34-050fe7af9700.png)

readevalprint commented:
That could be overkill, it might be reasonable to make all outgoing transactions require two steps.

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,133 @@
+        );
+    }
+
+    function setDAOMsg(address sender, address token, uint value) onlyDAO {
```
setDAOMsg is already done for you by the `Application` base class :)

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,133 @@
+            stringState = "Failed";
+    }
+
+    function AccountingApp() {
```
If you don't initialize Application with the DAO address, this contract won't produce bytecode as it is abstract. 

Example https://github.com/aragon/aragon-core/blob/master/contracts/apps/bylaws/BylawsApp.sol#L58

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,133 @@
+
+    // Create a new transaction and return the id of the new transaction.
+    // externalAddress is where the transication is coming or going to.
+    bytes4 constant NEW_TRANSACTION_SID = bytes4(sha3('newTransaction(address token, int value, address externalAddress, string reference, TransactionState initialState)'));
```
I like this right above the function.

Minor detail function signatures are calculated without the parameter names `sha3('newTransaction(address,int..`

izqui commented on contracts/apps/Application.sol:
```diff
@@ -1,6 +1,7 @@
 pragma solidity ^0.4.11;
 
 import "../dao/DAOStorage.sol";
+import "../misc/Crontab.sol";
```
does this have any reason to be here?

readevalprint replied:
yes I am about to add the crontab `next()` function now

izqui commented on test/dao_apps.js:
```diff
@@ -57,6 +63,35 @@ contract('Applications', accounts => {
       assert.isTrue(await mockApp.didStuff(), 'should have done stuff')
     })
+
+    it('can create new transaction', async () => {
```
We should move all this to its own test file (`integration_accounting` or something like that) as `dao_apps.js` tests that you can install apps in the DAO and not one app in particular.

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,186 @@
+        bytes2 ct_day;
+        bytes2 ct_month;
+        bytes2 ct_weekday;
+        uint start;
```
uint64 for timestamps is more than enough

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,186 @@
+            externalAddress: externalAddress, 
+            token: token, 
+            value: value, 
+            // TODO: get base token and exchange rate from oracle 
```
base token should be gotten from the current accounting period base token, right?

readevalprint replied:
ahh yes, english is ambiguous at times, it meant: get base token and then get exchange from oracle

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,186 @@
+        timestamp = t.timestamp;
+        state = tu.state;
+        accountingPeriodId = t.accountingPeriodId;
+        if (state == TransactionState.Succeeded) {
```
I wouldn't return strings from the contract. Should really be a frontend concern.

izqui commented on contracts/apps/accounting/AccountingApp.sol:
```diff
@@ -0,0 +1,209 @@
+     function approveTransaction(uint transactionId, string reason) onlyDAO {
+        Transaction memory t = transactions[transactionId];
+        if (t._type == TransactionType.Withdrawal){
+            VaultOrgan(dao).transfer(t.token, t.externalAddress, t.amount);
```
For avoiding reentrancy the token transfer should be the last thing executed. 

Also, how is this making sure a withdrawal cannot happen more than once?

readevalprint replied:
Excellent point, I will make `updateTransaction` internal and all the `setTransactionX` functions will validate the existing state first. and `approveTransaction` will require the current state to be `pending`. Actually I was thinking of renaming it `approveWithdrawal` because that's a bit more descriptive.
